src/pudl/output/ferc1.py

```diff
@@ -1677,3 +1790,508 @@ def remove_intra_table_calculated_values(exploded: pd.DataFrame) -> pd.DataFrame
+            return json.dumps(
+                [
+                    {
+                        "name": row.xbrl_factoid,
+                        "weight": float(row.weight),
+                        "xbrl_factoid_original": row.xbrl_factoid_original,
+                        "source_tables": [row.source_table],
+                    }
+                    for row in df.itertuples()
+                ]
+            )
+
+        root_calcs: pd.DataFrame = (
+            self.leafy_meta.reset_index()
+            .groupby(["root_table", "root_xbrl_factoid"], as_index=False)
+            .apply(leafy_meta_to_calculations)
+        )
+        root_calcs.columns = ["root_table", "root_xbrl_factoid", "calculations"]
+        return root_calcs
+
+    @property
+    def table_names(self: Self) -> list[str]:
+        """Produce the list of tables involved in this explosion."""
+        return list(self.exploded_meta.reset_index()["table_name"].unique())
+
+    @staticmethod
+    def plot(graph: nx.DiGraph, table_names: list[str]) -> None:
+        """Visualize the calculation forest and its attributes."""
+        colors = ["red", "orange", "yellow", "green", "blue", "purple"]
+        color_map = {
+            table: color
+            for table, color in zip(table_names, colors[: len(table_names)])
+        }
+        node_color = [color_map[node.source_table] for node in graph.nodes]
+
+        pos = graphviz_layout(graph, prog="dot", args='-Grankdir="LR"')
+        nx.draw_networkx_nodes(graph, pos, node_color=node_color)
+        nx.draw_networkx_edges(graph, pos)
+        # The labels are currently unwieldy
+        # nx.draw_networkx_labels(nx_forest, pos)
+        # Use this to draw everything if/once labels are fixed
+        # nx.draw_networkx(nx_forest, pos, node_color=node_color)
+        plt.show()
+
+    def plot_full_digraph(self: Self) -> None:
+        """Visualize the unpruned DAG."""
+        self.plot(self.full_digraph, table_names=self.table_names)
+
+    def plot_seeded_digraph(self: Self) -> None:
+        """Visualize the pruned forest."""
+        self.plot(self.seeded_digraph, table_names=self.table_names)
+
+    def plot_forest(self: Self) -> None:
+        """Visualize the pruned forest."""
+        self.plot(self.forest, table_names=self.table_names)
+
+    def leafy_data(
+        self: Self, exploded_data: pd.DataFrame, value_col: str
+    ) -> pd.DataFrame:
+        """Use the calculation forest to prune the exploded dataframe.
+
+        - Drop all rows that don't correspond to either root or leaf facts.
+        - Verify that the reported root values can still be generated by calculations
+          that only refer to leaf values.
+        - Merge the leafy metadata onto the exploded data, keeping only those rows
+          which refer to the leaf facts.
+        - Use the leaf weights to adjust the reported data values, and then drop the
+          leaf weights.
```

Comment:
We aren't dropping the columns here -- eventually it'll be done by `enforce_schema()`

Reply:
can you update the docs to be more reflective of whats going on in here then?

Reply:
Okay, updated.

 Given the NA values that are showing up in the leafy data, should we debug the missing `_correction` records? Or do an inner merge rather than a left merge onto the metdataframe?